src/mr_allocator.rs

```diff
@@ -1,60 +1,592 @@
+            debug!("remove item {:?}", item);
+        },
+    );
+}
+
+/// Register extent memory region with default access flags
+pub(crate) fn register_extent_mr_default(
+    addr: *mut c_void,
+    size: usize,
+    arena_ind: u32,
+) -> Option<Item> {
+    let access = ibv_access_flags::IBV_ACCESS_LOCAL_WRITE
+        | ibv_access_flags::IBV_ACCESS_REMOTE_WRITE
+        | ibv_access_flags::IBV_ACCESS_REMOTE_READ
+        | ibv_access_flags::IBV_ACCESS_REMOTE_ATOMIC;
+    register_extent_mr(addr, size, arena_ind, access)
+}
+
+/// Register extent memory region
+#[allow(clippy::as_conversions)]
+pub(crate) fn register_extent_mr(
+    addr: *mut c_void,
+    size: usize,
+    arena_ind: u32,
+    access: ibv_access_flags,
+) -> Option<Item> {
+    assert_ne!(addr, ptr::null_mut());
+    debug!(
+        "reg_mr addr {}, size {}, arena_ind {}, access {:?}",
+        addr as usize, size, arena_ind, access
+    );
+    let guard = pin();
+    ARENA_PD_MAP.get(&arena_ind, &guard).map_or_else(
+        || {
+            error!("can not get pd from ARENA_PD_MAP");
+            None
+        },
+        |pd| {
+            RawMemoryRegion::register_from_pd(pd, addr.cast::<u8>(), size, access).map_or_else(
+                |err| {
+                    error!("RawMemoryRegion::register_from_pd failed {:?}", err);
+                    None
+                },
+                |raw_mr| {
+                    Some(Item {
+                        addr: addr as usize,
+                        len: size,
+                        raw_mr: Arc::new(raw_mr),
+                        arena_ind,
+                    })
+                },
+            )
+        },
+    )
 }
 
 #[cfg(test)]
 mod tests {
-    use crate::RdmaBuilder;
-    use std::{alloc::Layout, io};
+    use super::*;
+    use crate::{context::Context, RdmaBuilder};
+    use std::{alloc::Layout, io, thread};
+    use tikv_jemalloc_sys::MALLOCX_ALIGN;
 
     #[tokio::test]
-    async fn test1() -> io::Result<()> {
+    async fn alloc_mr_from_rdma() -> io::Result<()> {
         let rdma = RdmaBuilder::default()
             .set_port_num(1)
             .set_gid_index(1)
             .build()?;
         let mut mrs = vec![];
+        let layout = Layout::new::<[u8; 4096]>();
         for _ in 0_i32..2_i32 {
-            let mr = rdma.alloc_local_mr(Layout::new::<[u8; 4096]>())?;
+            let mr = rdma.alloc_local_mr(layout)?;
             mrs.push(mr);
         }
         Ok(())
     }
+
+    #[test]
+    fn alloc_mr_from_allocator() -> io::Result<()> {
+        let ctx = Arc::new(Context::open(None, 1, 1)?);
+        let pd = Arc::new(ctx.create_protection_domain()?);
+        let allocator = Arc::new(MrAllocator::new(pd));
+        let layout = Layout::new::<char>();
+        let lmr = allocator.alloc(&layout)?;
+        debug!("lmr info :{:?}", &lmr);
+        Ok(())
+    }
+
+    #[test]
+    fn test_extent_hooks() -> io::Result<()> {
+        tracing_subscriber::fmt::init();
```

Comment:
There's no `assert` checking, I guess the test log output is the only output. We should create a customed hook for test, in which we insert some assertions.

Reply:
Yes, just some debug msg for people to observe the process of mr management.
So it is not a real test, and I will improve it later.